zerver/views/registration.py

```diff
@@ -137,9 +137,16 @@ def check_prereg_key(request: HttpRequest, confirmation_key: str) -> Preregistra
 
     prereg_user = get_object_from_key(confirmation_key, confirmation_types, activate_object=False)
 
-    if prereg_user.status == confirmation_settings.STATUS_REVOKED:
+    if prereg_user.status in [
+        confirmation_settings.STATUS_REVOKED,
+        confirmation_settings.STATUS_ACTIVE,
```

Comment:
Do we want a comment that `STATUS_ACTIVE` would previously be set for links that had been visited but not used, but that we expect that in practice all links with that property will already expired, since we made that change before adding configurable invitation link durations?

@alexmv FYI to confirm my understanding here. I guess I it might be better not as a comment, and instead being in the commit message, since it's irrelevant in the present codebase.

Reply:
I agree with the idea of renaming STATUS_ACTIVE to STATUS_USED as part of making this part of the codebase make sense. (The `status` field just didn't make sense in the original third-party confirmation library we imported circa 2012.)

That multi-use link issue seems worth discussing in chat.zulip.org... I am not sure that multi-use objects use the `status` field for anything, and feel like that'd be a better place to investigate together.

Reply:
@timabbott @alexmv I pushed a commit to rename to STATUS_USED.

As for multiuse invites, they don't actually have the `status` attribute:
```
class MultiuseInvite(models.Model):
    id: int = models.AutoField(auto_created=True, primary_key=True, verbose_name="ID")
    referred_by: UserProfile = models.ForeignKey(UserProfile, on_delete=CASCADE)
    streams: Manager = models.ManyToManyField("Stream")
    realm: Realm = models.ForeignKey(Realm, on_delete=CASCADE)
    invited_as: int = models.PositiveSmallIntegerField(
        default=PreregistrationUser.INVITE_AS["MEMBER"]
    )
```

This looks to me like it's all by design, because `get_object_from_key` has the `hasattr` part in `if activate_object and hasattr(obj, "status"):` probably exactly for this reason.

However I do think that this is all quite messy and confusing  :disappointed:

Reply:
I think setting `STATUS_USED` indeed doesn't make sense for multiuse invite links, but possibly we do want to support STATUS_REVOKED on them -- so maybe the right idea is that we should add a `status` column there, and make sure we don't pass the `activate_object` flag in that case.

Also, let's rename `activate_object` to `mark_object_used`? I've added a couple commits to try to improve this.

Reply:
> I think setting `STATUS_USED` indeed doesn't make sense for multiuse invite links, but possibly we do want to support STATUS_REVOKED on them -- so maybe the right idea is that we should add a `status` column there, and make sure we don't pass the `activate_object` flag in that case.

@timabbott  Hmm, we currently revoke multiuse invitiations by deletion, like this:
```
def do_revoke_user_invite(prereg_user: PreregistrationUser) -> None:
    email = prereg_user.email
    realm = prereg_user.realm
    assert realm is not None

    # Delete both the confirmation objects and the prereg_user object.
    # TODO: Probably we actually want to set the confirmation objects
    # to a "revoked" status so that we can give the invited user a better
    # error message.
    content_type = ContentType.objects.get_for_model(PreregistrationUser)
    Confirmation.objects.filter(content_type=content_type, object_id=prereg_user.id).delete()
    prereg_user.delete()
    clear_scheduled_invitation_emails(email)
    notify_invites_changed(realm)
```

So is the idea to replace that with setting `.status` on the MultiUseInvite to revoked? (this would also let us delete the `TODO` in the comment in this code)

Reply:
Yeah, exactly, setting it to REVOKED is what I had in mind...